### PR TITLE
Display used values as transparent

### PR DESCRIPTION
**Pull request by yanchith**
The feature is published as a check box in the main menu. Enabled by default,
now all used values/geometries are visible in the viewport, rendered with a
transparent(er) version of the currently selected viewport draw mode.

The renderer is refactored to divorce the concepts of a material and a viewport
draw mode. A viewport draw mode is what the application knows about and maps to
renderer materials. The renderer is now able to render each object within a
batch with a different material. Internally, a material is represented as a set
of values contained in a uniform buffer. Depending on a material's transparency
behavior (opaque/transparent/xray), the renderer automatically selects a correct
render pipeline to draw with - some pipelines read-write depth, some only read
depth, and some always pass the depth test without writing - and sorts the
objects in the three transparency category render lists (opaque, transparent,
xray) based on the object's centroid z value after projection with the camera
view matrix.

The session now notifies the caller with changes to used values. Internal value
diffing algorithm has been simplified, optimized and expanded to support
transitioning values between used and unused value sets.

The "Center viewport" feature now takes into consideration, whether used values
are visible, and always centers only around visible values.

Please test extensively. A lot of untested code has been changed and refactored.

### CODE REVIEW

**Review by janper — CHANGES_REQUESTED** [review]
The transparency level of shaded geometry is set way too high for the dark theme 
image

The wireframe zombie is indistinguishable form the live geometry 
image

The shaded zombie looks good in the funky theme 
image

The wireframe zombie is still bad even in the funky theme 
image

My proposal for the wireframe issue: can we render zombies as shaded without edges in all viewport modes?

Also there is a fundamental problem with the groups: however the group is created now (import or disjoint), it's impossible to do anything with it, but extract. When you extract, the group becomes a zombie an at the same place there is a new live geometry. Therefore you don't see the zombie group (sure, the group may be larger, so  you may see the outstanding bits). Only then you can do something with the mesh, say move it, which renders the extracted mesh a zombie, which is overlapping the zombie group. As a result, you'll almost never see a zombie group! My proposed solution: https://trello.com/c/35MgDAAI/303-for-purposes-of-the-mak-exhibition-replace-import-obj-as-group-with-import-obj-and-join-all-its-meshes-and-remove-both-extract-o

File: src/ui.rs [hunk 1]
```diff
@@ -609,6 +629,33 @@ impl<'a> UiFrame<'a> {
...
+                    viewport_draw_used_values,
+                );
+                if status.viewport_draw_used_values_changed {
+                    notifications.push(
+                        current_time,
+                        NotificationLevel::Info,
+                        if *viewport_draw_used_values {
+                            "Viewport now draws used geometry."
+                        } else {
+                            "Viewport now doesn't draw used geomery."
```
Comment (janper) [inline src/ui.rs]: geometry

File: src/ui.rs [hunk 2]
```diff
@@ -609,6 +629,33 @@ impl<'a> UiFrame<'a> {
...
+                        }
+                    );
+                }
+                if ui.is_item_hovered() {
+                    ui.tooltip(|| {
+                        let wrap_token = ui.push_text_wrap_pos(WRAP_POS_TOOLTIP_TEXT_PIXELS);
+                        ui.text("DRAW USED GEOMETRY IN VIEWPORT\n\
+                        \n\
+                        Used geometry is geometry that has been referenced as a parameter \
```
Comment (janper) [inline src/ui.rs:704]:
```
DRAW USED GEOMETRY IN VIEWPORT\n\
                        \n\
                        When enabled, used geometry will be drawn with \
                        a transparent material.\n\
                        \n\
                        Used geometry is geometry that has been referenced as a parameter \
                        in an operation. 
```

**Review by janper — CHANGES_REQUESTED** [review]

**Comment by yanchith** [comment]
> The transparency level of shaded geometry is set way too high for the dark theme

Ok, I'll make it less transparent for the dark theme.

> The wireframe zombie is indistinguishable form the live geometry

Yes, it was not meant to be different. I can not change it cheaply. This will be addressable in the future, though.

> The shaded zombie looks good in the funky theme

Funnily enough, this was certainly not intentional on my part. I think I do have a blending bug somewhere. Not sure I can fix it quickly, so I'll just pick the transparency value based on the selected theme and make sure it looks good on both.

> The wireframe zombie is still bad even in the funky theme

Same as in dark theme. Can't fix this quickly.

> My proposal for the wireframe issue: can we render zombies as shaded without edges in all viewport modes?

Yes, I'll do that.

> Also there is a fundamental problem with the groups: however the group is created now (import or disjoint), it's impossible to do anything with it, but extract. When you extract, the group becomes a zombie an at the same place there is a new live geometry. Therefore you don't see the zombie group (sure, the group may be larger, so you may see the outstanding bits). Only then you can do something with the mesh, say move it, which renders the extracted mesh a zombie, which is overlapping the zombie group. As a result, you'll almost never see a zombie group! My proposed solution: 

This can happen for non-group values as well (imagine transform with default parameters). This is correct behavior, not a defect. I'd do nothing to change this, both for MAK, and for later.

**Comment by janper** [comment]
> > Also there is a fundamental problem with the groups: however the group is created now (import or disjoint), it's impossible to do anything with it, but extract. When you extract, the group becomes a zombie an at the same place there is a new live geometry. Therefore you don't see the zombie group (sure, the group may be larger, so you may see the outstanding bits). Only then you can do something with the mesh, say move it, which renders the extracted mesh a zombie, which is overlapping the zombie group. As a result, you'll almost never see a zombie group! My proposed solution:
> 
> This can happen for non-group values as well (imagine transform with default parameters). This is correct behavior, not a defect. I'd do nothing to change this, both for MAK, and for later.

The problem is, that with the current toolset there is no possibility for seeing a zombie group. It simply can't happen, except for the rare cases when the group actually contains more than one geometry but we haven't come across such case with the hundreds of sample models we've got from SCD.

File: src/session.rs [hunk 3]
```diff
@@ -488,6 +431,179 @@ impl Session {
...
+    /// We diff the old and new sets and perform the following operations:
+    ///
+    /// - Detect which values should be added and add them (for both used and unused sets),
+    /// - Detect which values should be removed and remove them (for both used
+    ///   and unused sets),
+    /// - Detect which values should be retained, but they changed, we both
+    ///   remove and add them to notify the outside of the change (for both used
+    ///   and unused sets),
+    /// - Detect which values should be transitioned from the used set to unused
+    ///   and miami vice versa.
```
Comment (ondrowan) [inline src/session.rs]: Seriously? 😄

File: src/ui.rs [hunk 4]
```diff
@@ -40,6 +39,16 @@ pub enum Theme {
     Funky,
 }
 
+/// The draw mode applied to a group of objects in the viewport. Not always a
+/// 1:1 mapping with renderer materials.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum ViewportDrawMode {
```
Comment (ondrowan) [inline src/ui.rs:45]: This doesn't need to be `Eq` or `Hash`.

**Review by ondrowan — DISMISSED** [review]
I give this a quick look and found just 2 tiny things.

**Review by janper — CHANGES_REQUESTED** [review]
1. Would it be possible to disable shadows for zombies?

2. Came across this (see the shadows) 
without zombies
image

with zombies
image

**Comment by janper** [comment]
I see the self-transparency issue now 
image
nevermind, i can live with it

**Comment by yanchith** [comment]
> Would it be possible to disable shadows for zombies?

Yes. I'll do so.

> Came across this (see the shadows)

This probably wasn't introduced in this branch. It seems to be related to how we compute the bounding box that for what gets rendered into the shadow map. I'll create an issue in trello to solve that.

File: src/session.rs (hunk 3 above)
Comment (yanchith) [reply src/session.rs]: https://thumbs.gfycat.com/RemarkableTediousDwarfmongoose-small.gif

File: src/ui.rs (hunk 4 above)
Comment (yanchith) [reply src/ui.rs:45]: Hmmm, I can remove `Hash`, but I usually prefer to keep `Eq` to state future intentions, that this enum will always be `Eq`.

**Review by janper — APPROVED** [review]

**Comment by janper** [comment]
Absolutely perfect now! And it also makes sense and feels intuitive. Thank you!

**Pull request merged**